### DISCUSSION TITLE
Some magic needed to prevent overriding generic peripherals with dynamic peripherals

### DISCUSSION THREAD

**Discussion by SirEndii** [cc-tweaked/CC-Tweaked#1242] [Q&A] [answered]
Heyo
I currently have the issues that any `GenericPeripheral` is overwritten by our integration system.
As example, the basin from the create mod. ⇒ https://github.com/SirEndii/AdvancedPeripherals/issues/365
But we're unable to change this behavior without making too many changes.
We currently use `IDynamicPeripheral` for our [integration peripherals](https://github.com/SirEndii/AdvancedPeripherals/blob/1.19/src/main/java/de/srendi/advancedperipherals/lib/peripherals/IntegrationPeripheral.java#L17) to support our plugin system.

Is there any magic in cc that we can use to prevent this?

**Comment by SquidDev** [comment, marked as answer]
> Is there any magic in cc that we can use to prevent this?


Nooooo, not really.

**Reply by SirEndii** [reply]
Alrighty
Then I'll write peripheral plugins for the common capabilities so we just add these methods ourself

**Reply by SirEdvin** [reply]
Or just make ALL peripherals plugin-based :) https://github.com/cc-tweaked/CC-Tweaked/discussions/1093

**Reply by SirEndii** [reply]
This is an interesting approach
but I doubt that this will be reality in the near future

**Reply by SquidDev** [reply]
The design of the generic peripheral system was very much me looking at how complex Plethora's implementation was and going "hell no, let's make this simple". Now, I don't think I was successful at that - making things simple has never been my forte - but also I feel a UPW-style system takes things further towards Plethora's "hell no", not further away.

There's a lot of deficiencies in the generic peripheral system (I've bumped into a few recently), so I definitely appreciate any and all ideas on how to change/fix things. I sure don't trust myself to get it right :D.

**Reply by SquidDev** [reply]
This I guess also goes for the whole `@LuaFunction` system in general. One of its big issues is that everything is tied to the Java object hierarchy, which makes building anything which doesn't fit that model pretty awkward <sub><sup>(something something composition over inheritance)</sup></sub>. [We've actually got a hack around that](https://github.com/cc-tweaked/CC-Tweaked/blob/mc-1.19.x/projects/core/src/main/java/dan200/computercraft/core/asm/ObjectSource.java), it's just never made its way into the public API :D.

**Reply by SirEdvin** [reply]
I can probably suggest just move generic peripheral into public API part, with some ajustments (like adding extra itemHandler extractors, etc) and then just provide example of how another peripherals can just embed this peripheral into new one with IDynamicPeripheral magic

**Reply by SirEdvin** [reply]
Even providing extra class, that will handle this dynamic magic internally
